src/Package/Impl/DataInspect/Viewers/DebugObjectEvaluator.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.R.Debugger;
+using Microsoft.R.Host.Client;
+using Microsoft.VisualStudio.R.Package.Repl;
+
+namespace Microsoft.VisualStudio.R.Package.DataInspect.Viewers {
+    [Export(typeof(IDebugObjectEvaluator))]
+    public sealed class DebugObjectEvaluator : IDebugObjectEvaluator {
+        private readonly IDebugSessionProvider _debugSessionProvider;
+        private readonly IRSessionProvider _rSessionProvider;
+        private DebugSession _debugSession;
+        private IRSession _rSession;
+
+        [ImportingConstructor]
+        public DebugObjectEvaluator(IRSessionProvider rSessionProvider, IDebugSessionProvider debugSessionProvider) {
+            _rSessionProvider = rSessionProvider;
+            _debugSessionProvider = debugSessionProvider;
+        }
+
+        public async Task<DebugEvaluationResult> EvaluateAsync(string expression, DebugEvaluationResultFields fields) {
+            await TaskUtilities.SwitchToBackgroundThread();
+            DebugEvaluationResult result = null;
+
+            var debugSession = await GetDebugSessionAsync();
+            var frames = await debugSession.GetStackFramesAsync();
+
+            for (int i = frames.Count - 1; i >= 0; i--) {
```

Comment:
What is this loop trying to achieve? Find the correct frame to evaluate it in?

If so, that's not the right approach - the same expression can be legitimate in a different frame, but produce a totally different result (e.g. if it has a variable with the same name, but different value). It's why `DebugSession.EvaluateAsync` takes environment as one of the arguments. We should be doing the same thing here - capture the environment context, and pass it in.